### ISSUE TITLE
LSP positions need bound checks

### ISSUE THREAD

**Issue by wagnerf42** [helix-editor/helix#107]
hi,

in the following file:
```rust
struct Foo {}

impl std::ops::Drop for Foo {
    fn drop(&mut self) {
        println!("hello");
    }
}

fn foo() {
    let f = Foo {};
    panic!();
}
fn main() {
    foo()
}
```

deleting more than half of lines crashes.

i do 8xd and :

```
thread 'main' panicked at 'Attempt to index past end of Rope: line index 9, Rope line length 8', /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/ropey-1.2.0/src/rope.rs:946:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
   1: std::panicking::begin_panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:435:5
   2: ropey::rope::Rope::line_to_char
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/ropey-1.2.0/src/rope.rs:946:9
   3: helix_lsp::util::lsp_pos_to_pos
             at /home/wagnerf/code/helix/helix-lsp/src/lib.rs:58:28
   4: hx::application::Application::handle_language_server_message::{{closure}}::{{closure}}
             at /home/wagnerf/code/helix/helix-term/src/application.rs:179:49
   5: core::iter::adapters::map::map_fold::{{closure}}
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/map.rs:80:28
   6: core::iter::traits::iterator::Iterator::fold
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:2124:21
   7: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/map.rs:120:9
   8: core::iter::traits::iterator::Iterator::for_each
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:776:9
   9: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/spec_extend.rs:42:17
  10: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/spec_from_iter_nested.rs:53:9
  11: alloc::vec::source_iter_marker::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/source_iter_marker.rs:41:20
  12: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:2196:9
  13: core::iter::traits::iterator::Iterator::collect
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1765:9
  14: hx::application::Application::handle_language_server_message::{{closure}}
             at /home/wagnerf/code/helix/helix-term/src/application.rs:165:47
  15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  16: hx::application::Application::event_loop::{{closure}}
             at /home/wagnerf/code/helix/helix-term/src/application.rs:99:21
  17: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  18: hx::application::Application::run::{{closure}}
             at /home/wagnerf/code/helix/helix-term/src/application.rs:254:9
  19: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  20: hx::main::{{closure}}
             at /home/wagnerf/code/helix/helix-term/src/main.rs:169:5
  21: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  22: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/park/thread.rs:263:54
  23: tokio::coop::with_budget::{{closure}}
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:106:9
  24: std::thread::local::LocalKey<T>::try_with
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  25: std::thread::local::LocalKey<T>::with
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  26: tokio::coop::with_budget
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:99:5
  27: tokio::coop::budget
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/coop.rs:76:5
  28: tokio::park::thread::CachedParkThread::block_on
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/park/thread.rs:263:31
  29: tokio::runtime::enter::Enter::block_on
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/enter.rs:151:13
  30: tokio::runtime::thread_pool::ThreadPool::block_on
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/thread_pool/mod.rs:71:9
  31: tokio::runtime::Runtime::block_on
             at /home/wagnerf/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.6.1/src/runtime/mod.rs:452:43
  32: hx::main
             at /home/wagnerf/code/helix/helix-term/src/main.rs:171:5
  33: core::ops::function::FnOnce::call_once
             at /home/wagnerf/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

strangely i could not reproduce it with other files

**Comment by pickfire** [comment]
I got the same issue when doing https://github.com/helix-editor/helix/pull/125

**Comment by wagnerf42** [comment]
i don't think it is. i just pulled  and it is still here.

do you have rust-analyzer running for the language server ?  the pb comes from the warning on line 10.

**Comment by pickfire** [comment]
@wagnerf42 Can you please show the exact steps to reproduce? What did you mean by delete half the lines? Did you spam delete key or just select everything and delete? I tried having a whole document and spam ctrl-w in insert and it have no issue (except the last character being empty document but it was fixed).

**Comment by archseer** [comment]
I think this happens if you delete really quickly so that when (outdated) diagnostic messages come in, they no longer point to valid positions in the file. It's probably more prevalent on debug builds because the code is slower.

**Comment by wagnerf42** [comment]
> @wagnerf42 Can you please show the exact steps to reproduce? What did you mean by delete half the lines? Did you spam delete key or just select everything and delete? I tried having a whole document and spam ctrl-w in insert and it have no issue (except the last character being empty document but it was fixed).

i just open the file and hit '8xd' and it crashes.
i'm not in a debug build, i installed using cargo install.

**Comment by wagnerf42** [comment]
> I think this happens if you delete really quickly so that when (outdated) diagnostic messages come in, they no longer point to valid positions in the file. It's probably more prevalent on debug builds because the code is slower.

i think this is indeed the issue.

**Comment by wagnerf42** [comment]
so, you had me doubting and i did some more tests.

turns out after upgrading rust-analyzer the bug 'disappears'. (well, is not triggered anymore).

this was my previous version:
╭─wagnerf@grou ~  
╰─➤  rust-analyzer --version
rust-analyzer 1116c9a

so it might be possible to reproduce the bug by using an older rust-analyzer.

**Comment by pickfire** [comment]
Closing this since we can't find the method for reproduce. @wagnerf42 Thanks for reporting the issue.

**Comment by wagnerf42** [comment]
sorry i was not clear. here is the way to reproduce it:

install the old rust-analyzer:
git clone https://github.com/rust-analyzer/rust-analyzer.git && cd rust-analyzer
git checkout 1116c9a
cargo xtask install --server  

now with this version of rust-analyzer:
open the file
be on first line
8xd

**Comment by archseer** [comment]
I think I'm fine without an exact reproduction, we need to ensure the positions submitted by the LSP are in bounds or we risk panics.

**Comment by wojciechkepka** [comment]
> sorry i was not clear. here is the way to reproduce it:
> 
> install the old rust-analyzer:
> git clone https://github.com/rust-analyzer/rust-analyzer.git && cd rust-analyzer
> git checkout 1116c9a
> cargo xtask install --server
> 
> now with this version of rust-analyzer:
> open the file
> be on first line
> 8xd

I was able to reproduce the issue with this instructions, submitted a fix in #204 .
